File: wails_backend/internal/services/notifier.go
package services

import (
	"strconv"
	"strings"
)

// Notification setting keys and kinds.
//
// Settings are opt-out: a missing key means enabled, "0" disables it. Keep
// these in sync with the bun backend
// (bun_backend/server/services/notifier.ts) and the frontend defaults
// (frontend/lib/notifications.ts).
const (
	NotifyKeyEnabled  = "notifications.enabled"
	NotifyKeyReady    = "notifications.on_ready"
	NotifyKeyFinished = "notifications.on_finished"
	NotifyKeyError    = "notifications.on_error"
	NotifyKeyIdle     = "notifications.on_idle"
	NotifyKeyIdleSecs = "notifications.idle_seconds"

	DefaultIdleSeconds = 15
)

const (
	NotifyReady    = "ready"
	NotifyFinished = "finished"
	NotifyError    = "error"
	NotifyIdle     = "idle"
)

func notifyKey(kind string) string {
	switch kind {
	case NotifyReady:
		return NotifyKeyReady
	case NotifyFinished:
		return NotifyKeyFinished
	case NotifyError:
		return NotifyKeyError
	case NotifyIdle:
		return NotifyKeyIdle
	}
	return ""
}

func notifyEnabled(settings map[string]string, kind string) bool {
	if strings.TrimSpace(settings[NotifyKeyEnabled]) == "0" {
		return false
	}
	key := notifyKey(kind)
	return key != "" && strings.TrimSpace(settings[key]) != "0"
}

// idleSeconds is the delay before assuming a URL-less app is running.
func idleSeconds(settings map[string]string) int {
	n, err := strconv.Atoi(strings.TrimSpace(settings[NotifyKeyIdleSecs]))
	if err != nil || n < 1 {
		return DefaultIdleSeconds
	}
	if n > 600 {
		return 600
	}
	return n
}

File: bun_backend/server/services/notifier.ts
/**
 * Runner lifecycle notifications.
 *
 * Settings are stored in SQLite (Settings → Notifications). Everything is
 * opt-out: a missing key means enabled, "0" disables it. The same keys are
 * read by the Wails runner — keep them in sync with
 * `wails_backend/internal/services/notifier.go` and
 * `frontend/lib/notifications.ts`.
 */

import { settingsRepo } from "@db/settings"
import { notify } from "@native/notify"

export type NotificationKind = "ready" | "finished" | "error" | "idle"

export const NOTIFY_ENABLED_KEY = "notifications.enabled"
export const NOTIFY_IDLE_SECONDS_KEY = "notifications.idle_seconds"
export const DEFAULT_IDLE_SECONDS = 15

const KIND_KEYS: Record<NotificationKind, string> = {
  ready: "notifications.on_ready",
  finished: "notifications.on_finished",
  error: "notifications.on_error",
  idle: "notifications.on_idle",
}

function isEnabled(kind: NotificationKind): boolean {
  if (settingsRepo.get(NOTIFY_ENABLED_KEY) === "0") return false
  return settingsRepo.get(KIND_KEYS[kind]) !== "0"
}

/** Seconds to wait for a ready URL before assuming the app is running. */
export function idleSeconds(): number {
  const raw = Number(settingsRepo.get(NOTIFY_IDLE_SECONDS_KEY))
  if (!Number.isFinite(raw) || raw < 1) return DEFAULT_IDLE_SECONDS
  return Math.min(Math.round(raw), 600)
}

export function notifyApp(
  kind: NotificationKind,
  title: string,
  body?: string
): void {
  if (!isEnabled(kind)) return
  void notify({ title, body })
}
